libafl/src/stages/logics.rs

```diff
@@ -3,36 +3,31 @@
 use core::marker::PhantomData;
 
 use crate::{
-    stages::{HasCurrentStage, HasNestedStageStatus, Stage, StageProgress, StagesTuple},
+    stages::{HasCurrentStage, HasNestedStageStatus, Stage, StagesTuple},
     state::UsesState,
     Error,
 };
 
 /// Progress for nested stages. This merely enters/exits the inner stage's scope.
 #[derive(Debug)]
-pub struct NestedStageProgress;
+pub struct NestedStageRestartHelper;
 
-impl<S, ST> StageProgress<S, ST> for NestedStageProgress
-where
-    S: HasNestedStageStatus,
-{
-    fn initialize_progress(state: &mut S, _stage: &ST) -> Result<(), Error> {
+impl NestedStageRestartHelper {
+    fn restart_progress_should_run<S, ST>(state: &mut S, _stage: &ST) -> Result<bool, Error>
+    where
+        S: HasNestedStageStatus,
+    {
         state.enter_inner_stage()?;
-        Ok(())
+        Ok(true)
```

Comment:
this is not always true? right?
i think it is depending the result of the inner stage

Reply:
Don't think so

Reply:
hm okay i understand.

but i noticed one another problem.
These conditional stage evaluates the closure given by the user. but you could get stuck when evaluating them right? so how about also adding max # of retry to this helper too?

Reply:
ClosureStage uses the RetryHelper. For all user-provided stages, they'll have to implement their own restart handling

Reply:
I don't think bailing after `n` retries will always be the best solution (I would even say it's always the worst)